### PR TITLE
Enable mypy analysis for all files.

### LINKED ISSUE #573: Fix mypy type erorrs in skipped files
## Bug description
Some files are failing mypy type checks (but passing on pytype) and being skipped by `EXCLUDE_MYPY` in `.circleci/config.yml` introduced in https://github.com/HumanCompatibleAI/imitation/pull/534 We should fix these errors, by ignoring spurious errors and for true-positive errors refactoring the code or adding assert checks where appropriate.

## Steps to reproduce
Run `mypy` on the excluded files

### PR DESCRIPTION
**Pull request by ianyfan**
## Description

**EDIT: remaining issues have been resolved. The issues with `env.horizon` and `evaluate_policy` are to be taken upstream, and then depending on how quickly the turnaround is on those, either wait for them to be merged before merging this, or ignore the issues for now and un-ignore them once the fixes are done.**

ORIGINAL:

Fixes #573 by adding the remaining type annotations. Some of the issues in `mce_irl.py` is resolved by changing `env.horizon` to `int(env.horizon)` which indicates that there would be a problem if `env.horizon` was not an integer (I think it can be `np.inf`?). Does this occur in practice (since this should necessitate a runtime check)?

One other issue in `mce_irl.py` I fixed by using a `cast`, which I guarded behind `if TYPE_CHECKING:`. I don't think there is a better way of doing this, since there is no other way of restricting the type to what we want.

Remaining issues:
```
src/imitation/algorithms/preference_comparisons.py:121: error: Argument 1 to "shuffle" of "Generator" has incompatible type "List[TrajectoryWithRew]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"  [arg-type]
src/imitation/algorithms/preference_comparisons.py:644: error: Argument 1 to "choice" of "Generator" has incompatible type "List[TrajectoryWithRew]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"  [arg-type]
tests/algorithms/test_bc.py:133: error: Item "None" of "Optional[Cache]" has no attribute "mkdir"  [union-attr]
tests/algorithms/test_bc.py:163: error: Item "None" of "Optional[Cache]" has no attribute "mkdir"  [union-attr]
tests/algorithms/test_bc.py:186: error: Argument 1 to "evaluate_policy" has incompatible type "ActorCriticPolicy"; expected "BaseAlgorithm"  [arg-type]
tests/algorithms/test_bc.py:196: error: Argument 1 to "evaluate_policy" has incompatible type "ActorCriticPolicy"; expected "BaseAlgorithm"  [arg-type]
```

The ones in `preference_comparisons.py` appear to be due to an over-conservative annotation in `numpy`though I may have missed something. If it is due to this, then I'm not sure if there is a better way to deal with the issue than to `type: ignore` it.

The issues in `test_bc.py` are also weird and may just be worth ignoring.
`pytestconfig.cache` is typed as `Optional[Cache]` so I'm not sure whether this implies that it could actually be `None` sometimes, or whether it is just typed that way. If the latter, either an extra `assert is not None` may be sufficient; if the former, some extra logic may be required to run the tests properly.

## Testing

Ran `mypy` over relevant files, as well as `ci/code_checks.sh`.

### CODE REVIEW

File: src/imitation/algorithms/mce_irl.py
Comment (AdamGleave) [inline src/imitation/algorithms/mce_irl.py:59]: Can we add a TODO here so we remember to fix this before merge (if the plan is to fix it in a seals PR first), or open a GH issue to track it (if we want to merge this PR before fixing it upstream)?

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:1605]: as above would prefer to not change these types in this PR

File: src/imitation/data/rollout.py
Comment (AdamGleave) [inline src/imitation/data/rollout.py:429]: as above would prefer to not change these types in this PR

File: tests/algorithms/test_bc.py
Comment (AdamGleave) [inline tests/algorithms/test_bc.py:188]: Similar to with the seals upstream issue, let's add a TODO or GH issue for this.

File: src/imitation/algorithms/mce_irl.py
Comment (AdamGleave) [inline src/imitation/algorithms/mce_irl.py:397]: Copying from https://github.com/HumanCompatibleAI/imitation/pull/594/files (by @dan-pandori):
```suggestion
                # Safe since we know the first item is a Trajectory and `demonstrations`
                # is an Iterable
                demonstrations = cast(Iterable[types.Trajectory], demonstrations)
```

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:121]: From https://github.com/HumanCompatibleAI/imitation/pull/594/files
```suggestion
        # numpy's typing thinks it can't accept a list here, but it can
        self.rng.shuffle(trajectories)  # type: ignore[arg-type]
```

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:645]: From https://github.com/HumanCompatibleAI/imitation/pull/594/files
```suggestion
            # numpy's typing thinks it can't accept a list here, but it can
            traj = self.rng.choice(trajectories, p=p)  # type: ignore[arg-type]
```

File: tests/algorithms/test_bc.py
Comment (AdamGleave) [inline tests/algorithms/test_bc.py:130]: We use `pytestconfig.cache` a bunch of times here so I like the idea of adding a temporary variable `cache = pytestconfig.cache` from https://github.com/HumanCompatibleAI/imitation/pull/594/files#diff-648af1fe52e7f75f6ddc457b062fe9ee7b456d9358c493cef1a3547f5e87785b -- consider cherry-picking that change?

**Review by AdamGleave — CHANGES_REQUESTED** [review]
Largely looks good, only some minor changes needed.

**Comment by AdamGleave** [comment]
On upstream: I don't think we're going to have time to turnaround the SB3 change before you wrap up. The seals one should be a short fix, and I can review and merge same-day, so let's push for fixing that one upstream though.

File: src/imitation/algorithms/mce_irl.py
Comment (ianyfan) [reply src/imitation/algorithms/mce_irl.py]: I've opened a PR and added an extra check that should still be relevant even once the upstream has been changed, but can still add an extra comment if that would be helpful

File: src/imitation/algorithms/dagger.py
Comment (ianyfan) [reply src/imitation/algorithms/dagger.py:165]: Ah it happens, at least this appears to be the only occurrence. Though to be honest, I thought this sort of thing usually gets picked up by linters, maybe that can be added

File: src/imitation/algorithms/mce_irl.py
Comment (ianyfan) [reply src/imitation/algorithms/mce_irl.py:413]: I've found a better way to do this, I think

**Comment by ianyfan** [comment]
There's more places that are missing type annotations, but since they don't cause errors in mypy, I will leave them be, so as not to add any more noise to this PR.

**Review by AdamGleave — APPROVED** [review]
LGTM.

@Rocamonde, I made some commits to this PR, can you please review the diff: https://github.com/HumanCompatibleAI/imitation/pull/593/files/f3805f8c7e2af8dd4e3803325ae586dee9e6f7ef..b89b7f067f04576b7845ead486089e46317f19a6

**Comment by ianyfan** [comment]
New changes look good to me, at least

**Comment by Rocamonde** [comment]
Apologies, this did not show up on my GitHub PR notifications. Will review tomorrow.

File: src/imitation/algorithms/mce_irl.py
Comment (Rocamonde) [inline src/imitation/algorithms/mce_irl.py:509]: Before, the list comprehension added the gradient of the parameter, `p.grad`. Now you're adding the parameter, `p`. This is probably a bug?

File: tests/algorithms/test_bc.py
Comment (Rocamonde) [inline tests/algorithms/test_bc.py:189]: May I suggest
```suggestion
    # TODO(https://github.com/HumanCompatibleAI/imitation/issues/600): the upstream annotation for this function is overly-conservative
```
Easier to click?

File: tests/algorithms/test_mce_irl.py
Comment (Rocamonde) [inline tests/algorithms/test_mce_irl.py:345]: Nice! Cool to see the wrapper being used in a composable, clean way in tests.

**Review by Rocamonde — CHANGES_REQUESTED** [review]
Looks good, just one thing I noticed that might be a bug, do take a look.

**Comment by Rocamonde** [comment]
Yay! Thanks for doing this @ianyfan , nice to see what we started being nicely completed. This will certainly make our codebase more robust to random bugs and errors.

File: src/imitation/algorithms/mce_irl.py
Comment (AdamGleave) [reply src/imitation/algorithms/mce_irl.py:509]: Oh dear, yes this is a bug, thanks for catching! It's a little sad this was not caught by our tests. :/

File: tests/algorithms/test_bc.py
Comment (AdamGleave) [reply tests/algorithms/test_bc.py:189]: Makes line too long but it's a multi-line comment so sure I'll just reflow it.

**Comment by AdamGleave** [comment]
Thanks for the review @Rocamonde, issues should be resolved now

File: src/imitation/algorithms/density.py
Comment (Rocamonde) [inline src/imitation/algorithms/density.py:346]: We could try to be more specific than this, but ParamSpec is not (yet) a very well supported feature by most type checkers, and it might not be worth the effort of typing it.

File: src/imitation/algorithms/mce_irl.py
Comment (Rocamonde) [inline src/imitation/algorithms/mce_irl.py:195]: Interesting . Is this type not incompatible with the superclass? How does that work?

File: src/imitation/algorithms/mce_irl.py
Comment (Rocamonde) [inline src/imitation/algorithms/mce_irl.py:400]: I'm not sure how this type is incompatible since we are narrowing the scope of the type, we should be able to use the same variable if the generic is well constructed. (right?)

File: src/imitation/algorithms/mce_irl.py
Comment (Rocamonde) [inline src/imitation/algorithms/mce_irl.py:506]: I guess you could also just cast it and keep it was it was. But I don't really feel strongly about it, how it's now seems fine too. Just not obvious why we are doing it, so maybe add a comment saying it's for the type checker.

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [inline src/imitation/algorithms/preference_comparisons.py:645]: same comment as above

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [inline src/imitation/algorithms/preference_comparisons.py:910]: Is this the google style guide?

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [inline src/imitation/algorithms/preference_comparisons.py:972]: `i` can also be a slice object in general. Is there a reason why we're not allowing this here? (also, you would have to change the return type through an overload)

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [inline src/imitation/algorithms/preference_comparisons.py:1175]: thanks, i remember this being an issue when splitting the preference dataset (you get a subset type) but did not want to change the type as I was not sure that was fine.

File: src/imitation/rewards/serialize.py
Comment (Rocamonde) [inline src/imitation/rewards/serialize.py:177]: Maybe this is worth making positional only in the original type and everywhere else?

**Review by Rocamonde — COMMENTED** [review]
This is excellent work! LGTM. I left some comments with suggestions and questions, worth looking over them before merging to make sure we're happy with the direction of things.

File: src/imitation/algorithms/dagger.py
Comment (ianyfan) [reply src/imitation/algorithms/dagger.py:47]: https://mypy.readthedocs.io/en/stable/class_basics.html#annotating-init-methods I think it is an error to not return None. However, omitting it would not cause an error to be thrown.

This has not been added everywhere. In fact, there are a large number of missing annotations. I started to go through them, but stopped because this PR is in principle just to fix the errors that were raised by enabling mypy on these files, and I didn't want to add lots of extra noise to the changes. However, I kept the changes in the commits I had already pushed.

File: src/imitation/algorithms/mce_irl.py
Comment (ianyfan) [reply src/imitation/algorithms/mce_irl.py:400]: I presume mypy doesn't actually narrow the scope of the variable, even if you use it like this. Not sure though

File: src/imitation/algorithms/preference_comparisons.py
Comment (ianyfan) [reply src/imitation/algorithms/preference_comparisons.py:910]: I'm not sure. It's not a very important change anyway, I think, so can just revert it

File: setup.py
Comment (AdamGleave) [reply setup.py:201]: Yeah, I agree. We decided in another PR to use `==` for test-only dependencies that had `0.x` revisions because semantic versioning says they can make breaking API changes at any time. However, now that it's an install requirement not a test requirement we should be permissive with versions so as not to cause conflicts for our users. With seals in particular if we make a breaking change there and forget to update `imitation` I have limited sympathy for us!

I will note that we'll still need to bump the version in `setup.py` for any changes that are _needed_ in imitation: CI doesn't install new versions automatically, only when `setup.py` is changed invalidating the existing cache.

File: src/imitation/algorithms/dagger.py
Comment (AdamGleave) [reply src/imitation/algorithms/dagger.py:47]: Adding `-> None` to `__init__` seems like it should be doable with a `sed` one-liner. But I agree keeping this PR minimal is good, if we have another PR that just does that change it'll be easy to review because all the changes will be the same...

Having `__init__` annotated with None feels kinda silly to me: what else is it going to return? But according to https://stackoverflow.com/a/73334605/14507224 `mypy` will complain if there is an `__init__` method with no type annotations (it's OK if one of the arguments is typed though), so there's some argument for adding it everywhere for consistency.

Comment (AdamGleave) [reply src/imitation/algorithms/dagger.py:47]: OK I change my mind, adding `-> None` to `__init__` does seem good given https://peps.python.org/pep-0484/ says:
> (Note that the return type of __init__ ought to be annotated with -> None. The reason for this is subtle. If __init__ assumed a return annotation of -> None, would that mean that an argument-less, un-annotated __init__ method should still be type-checked? Rather than leaving this ambiguous or introducing an exception to the exception, we simply say that __init__ ought to have a return annotation; the default behavior is thus the same as for other methods.)

File: src/imitation/algorithms/density.py
Comment (AdamGleave) [reply src/imitation/algorithms/density.py:346]: I've not used `ParamSpec` much so may be missing something, but how would it help us here? It seems designed in cases where the `P = ParamSpec("P")` variable is reused, like decorators or other higher-order functions. But if we're returning `None` I don't think it makes a difference.

File: src/imitation/algorithms/mce_irl.py
Comment (AdamGleave) [reply src/imitation/algorithms/mce_irl.py:195]: I think it works because `NoReturn` is the bottom type: i.e. a type that is a subtype of all other types. There's a good discussion at https://github.com/python/mypy/issues/4116

File: src/imitation/algorithms/mce_irl.py
Comment (AdamGleave) [reply src/imitation/algorithms/mce_irl.py:506]: Gradients really can be none in general, but it's true we know they're not here because we called `loss.backward()` in the `self._train_step` call, so a cast does seem fine. I've added a comment for now, but happy to change it to cast, I was pretty on the fence about it.

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:641]: It looks like `numpy` is fine with lists actually, the issue is just that it expects the arrays to be of scalars:

```
(venv) adam@puffin:~/dev/imitation$ cat test.py
from numpy import random

rng : random.Generator = random.default_rng()

my_list = [1, 2, 3, 4, 5]
random.shuffle(my_list)
rng.shuffle(my_list)

class Foo:
    def __init__(self, x) -> None:
        self.x = x

my_complex_list = [Foo(42), Foo(21)]
random.shuffle(my_complex_list)
rng.shuffle(my_complex_list)
(venv) adam@puffin:~/dev/imitation$ mypy test.py
test.py:14: error: Argument 1 has incompatible type "List[Foo]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"
test.py:15: error: Argument 1 to "shuffle" of "Generator" has incompatible type "List[Foo]"; expected "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]"
Found 2 errors in 1 file (checked 1 source file)
```

I'm not sure whether they'd really want to change it given this, the one for `shuffle` is already maximally permissive within their type hierarchy taking `ArrayLike`: https://github.com/numpy/numpy/blob/266aad7478bc7fbcc55eea7f942a0d373b838396/numpy/random/_generator.pyi But `ArrayLike` is just quite restrictive: https://github.com/numpy/numpy/blob/main/numpy/_typing/_array_like.py#L85  Seems like a deliberate choice, changing that might break a lot of other annotations...

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:906]: We were already importing this as `data_th` so it seems good to do this for consistency IMO.

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:968]: Good idea, added!

File: src/imitation/algorithms/preference_comparisons.py
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:906]: I don't think it matches the signature for Sequence, `__getitem__` with a slice does not return a Sequence of tuples but a tuple of sequences. We could handle int and slice keys differently like in types, but reluctant to add that complexity until we have a use case for it.

File: src/imitation/rewards/serialize.py
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py:177]: I'm always a bit scared of positional functions where all arguments have the same type, very easy to get them mixed up. I'd rather just actually stick to a convention here, the current one seems OK.

**Comment by AdamGleave** [comment]
@Rocamonde I believe I've addressed all outstanding comments on this, please do re-review when you have a moment. Note we'll need to merge this one before https://github.com/HumanCompatibleAI/imitation/pull/603 because that removes the mypy excludes.

**Comment by Rocamonde** [comment]
Sorry about this. I reviewed the PRs sequentially from my inbox and only saw this should have been merged before now.

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [reply src/imitation/algorithms/preference_comparisons.py:907]: Good catch, makes sense.

Comment (Rocamonde) [reply src/imitation/algorithms/preference_comparisons.py:907]: Cool.

File: src/imitation/algorithms/mce_irl.py
Comment (Rocamonde) [reply src/imitation/algorithms/mce_irl.py:506]: Yeah makes sense. I think the assert is slightly safer, and if there's a comment it should be OK.

File: src/imitation/algorithms/preference_comparisons.py
Comment (Rocamonde) [reply src/imitation/algorithms/preference_comparisons.py:642]: Yeah I see, interesting, let's just deal with this internally then.

File: src/imitation/algorithms/density.py
Comment (Rocamonde) [reply src/imitation/algorithms/density.py:346]: I meant for Any, not for None haha. The problem with this is that any subtype has to take any kwarg in the signature, but you only really want it to take the kwargs that `.learn()` can take. But not sure how you can use ParamSpec for this now I think about it. I think it was designed as a generic that is filled in implicitly at the callsite, not for extracting the type signature of an explicit method (which tbh should be a valid use case).

**Comment by Rocamonde** [comment]
Actually it was fine, the merge was trivial. Will wait for tests to run and then merge.

**Review by Rocamonde — APPROVED** [review]
LGTM

**Pull request merged**
